matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/task/ConfigurableTask.kt

```diff
@@ -16,57 +16,67 @@
+internal fun <PARAMS, RESULT> Task<PARAMS, RESULT>.configureWith(params: PARAMS,
+                                                                 init: (ConfigurableTask.Builder<PARAMS, RESULT>.() -> Unit) = {}
+): ConfigurableTask<PARAMS, RESULT> {
+    return ConfigurableTask.Builder(this, params).apply(init).build()
 }
 
-/**
- * Convert a Task to a ConfigurableTask without parameter
- */
-internal fun <RESULT> Task<Unit, RESULT>.toConfigurableTask(): ConfigurableTask<Unit, RESULT> {
-    return ConfigurableTask(this, Unit)
+internal fun <RESULT> Task<Unit, RESULT>.configureWith(init: (ConfigurableTask.Builder<Unit, RESULT>.() -> Unit) = {}): ConfigurableTask<Unit, RESULT> {
+    return configureWith(Unit, init)
 }
 
 internal data class ConfigurableTask<PARAMS, RESULT>(
         val task: Task<PARAMS, RESULT>,
         val params: PARAMS,
-        val callbackThread: TaskThread = TaskThread.MAIN,
-        val executionThread: TaskThread = TaskThread.IO,
-        val retryCount: Int = 0,
-        val callback: MatrixCallback<RESULT> = object : MatrixCallback<RESULT> {}
-) : Task<PARAMS, RESULT> {
-
+        val id: UUID,
+        val callbackThread: TaskThread,
+        val executionThread: TaskThread,
+        val constraints: TaskConstraints,
+        val retryCount: Int,
+        val callback: MatrixCallback<RESULT>
 
-    override suspend fun execute(params: PARAMS): Try<RESULT> {
-        return task.execute(params)
-    }
+) : Task<PARAMS, RESULT> {
 
-    fun callbackOn(thread: TaskThread): ConfigurableTask<PARAMS, RESULT> {
-        return copy(callbackThread = thread)
-    }
 
-    fun executeOn(thread: TaskThread): ConfigurableTask<PARAMS, RESULT> {
-        return copy(executionThread = thread)
+    class Builder<PARAMS, RESULT>(
+            private val task: Task<PARAMS, RESULT>,
+            private val params: PARAMS,
+            var id: UUID = UUID.randomUUID(),
+            var callbackThread: TaskThread = TaskThread.MAIN,
+            var executionThread: TaskThread = TaskThread.IO,
+            var constraints: TaskConstraints = TaskConstraints(),
+            var retryCount: Int = 0,
+            var callback: MatrixCallback<RESULT> = object : MatrixCallback<RESULT> {}
+    ) {
+
+        fun build() = ConfigurableTask(
+                task = task,
+                params = params,
+                id = id,
+                callbackThread = callbackThread,
+                executionThread = executionThread,
+                constraints = constraints,
+                retryCount = retryCount,
+                callback = callback
+        )
     }
 
-    fun dispatchTo(matrixCallback: MatrixCallback<RESULT>): ConfigurableTask<PARAMS, RESULT> {
-        return copy(callback = matrixCallback)
-    }
 
-    fun enableRetry(retryCount: Int = Int.MAX_VALUE): ConfigurableTask<PARAMS, RESULT> {
-        return copy(retryCount = retryCount)
+    override suspend fun execute(params: PARAMS): RESULT {
```

Comment:
You can remove this if you change line 43 to `) : Task<PARAMS, RESULT> by task {`.